### PR TITLE
array-utilities

### PR DESCRIPTION
**Pull request by aerorohit**
Adds many conveniences for array objects:
- Adds `linspace`, `diff`, `trapz`,  `where`
- comparison operators for an array object
- tests for above functions and `argsort`, `ones_like`, `minimum`, `sum`
- removed unnecessary import of cffi to support older versions of pycuda.

### CODE REVIEW

**Comment by prabhuramachandran** [comment]
Thanks @aerorohit can you please update the message here on the PR to summarize all the key changes you have made?

**Comment by aerorohit** [comment]
Functions added in array.py

- `linspace`
- `diff`
- `trapz`
- `where`
- comparison operators for an array object
- tests for above functions and `argsort`, `ones_like`, `minimum`, `sum`
- removed unnecessary import of cffi to support older versions of pycuda.

File: compyle/array.py [hunk 1]
```diff
@@ -372,6 +368,85 @@ def arange(start, stop, step, dtype=np.int32, backend='cython'):
     return wrap_array(out, backend)
 
 
+def linspace(start, stop, num, dtype=np.float64, backend='opencl',
+             endpoint=True):
+    if not type(num) == int:
+        raise Exception("num should be an integer")
```
Comment (prabhuramachandran) [inline compyle/array.py:374]: Please provide a more helpful message here.

File: compyle/array.py [hunk 2]
```diff
@@ -372,6 +368,85 @@ def arange(start, stop, step, dtype=np.int32, backend='cython'):
...
+    if num <= 0:
+        raise Exception('num should be greater than 0')
```
Comment (prabhuramachandran) [inline compyle/array.py:376]: Same here.

File: compyle/array.py [hunk 3]
```diff
@@ -372,6 +368,85 @@ def arange(start, stop, step, dtype=np.int32, backend='cython'):
...
+        out = gpuarray.arange(0, num, 1, dtype=dtype)
+        out = out*delta+start
+    else:
+        out = np.linspace(start, stop, num,
+                          endpoint=endpoint, dtype=dtype)
+    return wrap_array(out, backend)
+
+
+@annotate
+def nDiff(i, y, x, b, lb):
```
Comment (prabhuramachandran) [inline compyle/array.py:403]: Can you change the function name to be consistent with the rest, i.e. use `lower_case_with_underscores`.

File: compyle/array.py [hunk 4]
```diff
@@ -372,6 +368,85 @@ def arange(start, stop, step, dtype=np.int32, backend='cython'):
...
+    """
+    if n == 0:
+        return a
+    if n < 0:
+        raise ValueError(
+            "order must be non-negative but got " + repr(n))
+    if(a.__len__() < n+1):
+        raise ValueError(
+            "Array a should have length at least n+1, but got "
+            + str(a.__len__()))
```
Comment (prabhuramachandran) [inline compyle/array.py:429]: Use `len(a)` and not `a.__len__()`.

File: compyle/array.py [hunk 5]
```diff
@@ -372,6 +368,85 @@ def arange(start, stop, step, dtype=np.int32, backend='cython'):
...
+    The first difference is given by ``out[i] = a[i+1] - a[i]`` along
+    the given axis, higher differences are calculated by using `diff`
+    recursively.
+    """
+    if n == 0:
+        return a
+    if n < 0:
+        raise ValueError(
+            "order must be non-negative but got " + repr(n))
+    if(a.__len__() < n+1):
```
Comment (prabhuramachandran) [inline compyle/array.py:426]: Do not call private methods, use `len(a)`

File: compyle/array.py [hunk 6]
```diff
@@ -372,6 +368,85 @@ def arange(start, stop, step, dtype=np.int32, backend='cython'):
...
+        backend = a.backend
+
+    if backend == 'opencl' or backend == 'cuda':
+        from compyle.api import Elementwise
+        binom_coeff = np.zeros(n+1)
+        sign_fac = 1 if (n % 2 == 0) else -1
+        for i in range(n+1):
+            binom_coeff[i] = choose(n, i) * (-1)**i * sign_fac
+        binom_coeff = wrap(binom_coeff, backend=backend)
+        len_ar = a.__len__()
```
Comment (prabhuramachandran) [inline compyle/array.py:441]: Not sure why you are using this, `__len__` please remove all such occurrences.

File: compyle/array.py [hunk 7]
```diff
@@ -424,6 +499,48 @@ def dot(a, b, backend=None):
...
 
 
+def trapz(y, x=None, dx=1.0, backend=None):
+    if backend is None:
+        backend = y.backend
+    if x is None:
+        d = dx
+        out = (sum(y, backend=backend) - 0.5 * (y[0] + y[-1])) * d
+    else:
+        if not x.__len__() == (y.__len__()):
```
Comment (prabhuramachandran) [inline compyle/array.py:509]: Here too!

File: compyle/array.py [hunk 8]
```diff
@@ -609,7 +726,7 @@ def out_cumsum(i, ary, out, item):
 def cumsum(ary, backend=None, out=None):
     if backend is None:
         backend = ary.backend
-    if backend == 'opencl' or backend == 'cuda':
+    if backend == 'opencl' or backend == 'cuda' or backend == 'C':
```
Comment (prabhuramachandran) [inline compyle/array.py:729]: What is this backend C?

File: compyle/array.py [hunk 9]
```diff
@@ -624,6 +741,22 @@ def cumsum(ary, backend=None, out=None):
...
+        out_ar[cum_sum_ar[i]-1] = ary[i]
+
+
+def take_bool(ary, condition, backend=None):
+    if backend is None:
+        backend = ary.backend
+    cumsum_ar = cumsum(condition, backend=backend)
+    out_ar = ones(cumsum_ar[-1], ary.dtype, backend=backend)
+    take_bool_elwise_an = annotate(take_bool_elwise)
+    e = Elementwise(take_bool_elwise_an, backend=backend)
```
Comment (prabhuramachandran) [inline compyle/array.py:755]: This should be cached right?  Each time `take_bool` is called it is going to create a fresh elementwise, you should use some form of memoization as done in many places in compyle.

File: compyle/array.py [hunk 10]
```diff
@@ -424,6 +499,48 @@ def dot(a, b, backend=None):
...
+                'x and y should have same backend, got ${x_bk} and ${y_bk}'.
+                format(x_bk=x.backend, y_bk=y.backend))
+    if x.dtype is not y.dtype:
+        raise TypeError(
+            'x and y should have same data type, got {} and {}'.format(
+                x.dtype, y.dtype))
+
+    xtype_c = 'g' + dtype_to_ctype(x.dtype) + 'p'
+    where_annotated = annotate(where_elwise)
+    e = Elementwise(where_annotated, backend=backend)
```
Comment (prabhuramachandran) [inline compyle/array.py:538]: This should be memoized.

File: compyle/array.py [hunk 11]
```diff
@@ -372,6 +368,85 @@ def arange(start, stop, step, dtype=np.int32, backend='cython'):
...
+    if backend == 'opencl' or backend == 'cuda':
+        from compyle.api import Elementwise
+        binom_coeff = np.zeros(n+1)
+        sign_fac = 1 if (n % 2 == 0) else -1
+        for i in range(n+1):
+            binom_coeff[i] = choose(n, i) * (-1)**i * sign_fac
+        binom_coeff = wrap(binom_coeff, backend=backend)
+        len_ar = a.__len__()
+        y = zeros(len_ar - n, dtype=a.dtype, backend=backend)
+        diff_kernel = Elementwise(nDiff, backend=backend)
```
Comment (prabhuramachandran) [inline compyle/array.py:443]: Memoization please.

File: compyle/array.py [hunk 12]
```diff
@@ -681,6 +814,54 @@ def align(ary_list, order, out_list=None, backend=None):
...
+def comparison_template(func, other, arr, backend=None):
+    if backend is None:
+        backend = arr.backend
+    from compyle.parallel import Elementwise
+    other_type = dtype_to_ctype(type(other))
+    ary_type = dtype_to_ctype(arr.dtype) + 'p'
+    ans = empty(arr.length, dtype=np.int32, backend=arr.backend)
+    func_annotated = annotate(func, i='int', x=ary_type,
+                              val=other_type, ans='intp')
+    e = Elementwise(func_annotated, backend=arr.backend)
```
Comment (prabhuramachandran) [inline compyle/array.py:850]: Memoization

File: compyle/array.py [hunk 13]
```diff
@@ -732,15 +917,21 @@ def __setitem__(self, key, value):
 
     def __add__(self, other):
         if isinstance(other, Array):
-            other = other.dev
-        ans = self.dev + other
-        return wrap_array(ans, self.backend)
+            e = Elementwise(add_elwise, backend=self.backend)
```
Comment (prabhuramachandran) [inline compyle/array.py:920]: Again here.

File: compyle/array.py (hunk 1 above)
Comment (prabhuramachandran) [reply compyle/array.py:374]: Also a better exception, not a generic Exception.

File: compyle/tests/test_array.py [hunk 14]
```diff
@@ -360,3 +383,191 @@ def test_cumsum(backend):
...
+        get_config().use_double = True
+    # Given
+    x = array.arange(0., 10., 1., dtype=dtype, backend=backend)
+
+    # When
+    out = x.__getattribute__(method)(5)
+
+    # Then
+    x_np = np.arange(10, dtype=dtype)
+    comp = [int(i) for i in x_np.__getattribute__(method)(5)]
```
Comment (prabhuramachandran) [inline compyle/tests/test_array.py:460]: Arghh,  never call these protected methods, use `getattr(x, method)`.

File: compyle/tests/test_array.py [hunk 15]
```diff
@@ -360,3 +383,191 @@ def test_cumsum(backend):
...
+@check_comparison_methods
+def test_comparison(backend, dtype, method):
+    check_import(backend)
+    if dtype == np.float64:
+        get_config().use_double = True
+    # Given
+    x = array.arange(0., 10., 1., dtype=dtype, backend=backend)
+
+    # When
+    out = x.__getattribute__(method)(5)
```
Comment (prabhuramachandran) [inline compyle/tests/test_array.py:456]: Never call these in public, use `getattr(x, method)`.

**Review by prabhuramachandran — CHANGES_REQUESTED** [review]
Please make the requested changes. Thanks.

File: compyle/array.py [hunk 16]
```diff
@@ -506,7 +513,7 @@ def trapz(y, x=None, dx=1.0, backend=None):
         d = dx
         out = (sum(y, backend=backend) - 0.5 * (y[0] + y[-1])) * d
     else:
-        if not x.__len__() == (y.__len__()):
+        if not len(x) == (len(y)):
```
Comment (prabhuramachandran) [inline compyle/array.py:516]: Remove extra brackets around `(len(y))`.

File: compyle/array.py [hunk 17]
```diff
@@ -521,6 +528,13 @@ def where_elwise(i, condition, x, y,  ans):
         ans[i] = y[i]
 
 
+@memoize
+def where_kernel(backend, dtype):
+    where_annotated = annotate(where_elwise)
```
Comment (prabhuramachandran) [inline compyle/array.py:533]: Can't you use the annotate decorator for the `where_elwise`?

**Review by prabhuramachandran — CHANGES_REQUESTED** [review]
A few more minor issues.

**Review by prabhuramachandran — APPROVED** [review]

**Pull request merged**